starfish/recipe/runnable.py

```diff
@@ -0,0 +1,254 @@
+            PipelineComponentType.get_pipeline_component_type_by_name(self._pipeline_component_name)
+        self._algorithm_cls: Type = getattr(
+            self._pipeline_component_cls, self._algorithm_name)
+
+        # retrieve the actual __init__ method
+        signature = Runnable._get_actual_method_signature(
+            self._algorithm_cls.__init__)
+        formatted_algorithm_options = self._format_algorithm_constructor_arguments(
+            signature, self._raw_algorithm_options, self.__str__)
+
+        try:
+            self._algorithm_instance: AlgorithmBase = self._algorithm_cls(
+                **formatted_algorithm_options)
+        except Exception as ex:
+            raise ConstructorError(f"Error instantiating the algorithm for {str(self)}") from ex
+
+        # retrieve the actual run method
+        signature = Runnable._get_actual_method_signature(
+            self._algorithm_instance.run)  # type: ignore
+        self._inputs = self._format_run_arguments(signature, self._raw_inputs, self.__str__)
+
+    @staticmethod
+    def _get_actual_method_signature(run_method: Callable) -> inspect.Signature:
+        if hasattr(run_method, "__closure__"):
+            # it's a closure, probably because of AlgorithmBaseType.run_with_logging.  Unwrap to
+            # find the underlying method.
+            closure = run_method.__closure__  # type: ignore
+            if closure is not None:
+                run_method = closure[0].cell_contents
+
+        return inspect.signature(run_method)
+
+    @staticmethod
+    def _format_algorithm_constructor_arguments(
+            constructor_signature: inspect.Signature,
+            algorithm_options: Mapping[str, Any],
+            str_callable: Callable[[], str],
+    ) -> Mapping[str, Any]:
+        """Given the constructor's signature and a mapping of keyword argument names to values,
+        format them such that the constructor can be invoked.
+
+        Some parameters may be :py:class:`starfish.recipe.filesystem.FileProvider` instances.  Use
+        the type hints in the constructor's signature to identify the expected file type, and load
+        them into memory accordingly.
+
+        Parameters
+        ----------
+        constructor_signature : inspect.Signature
+            The signature for the constructor.
+        algorithm_options : Mapping[str, Any]
+            The parameters for the constructor, as provided to the Runnable.
+        str_callable : Callable[[], str]
+            A callable that can be invoked to provide a user-friendly representation of the
+            Runnable, in case any errors or warnings are generated.
+
+        Returns
+        -------
+        Mapping[str, Any] : The parameters for the constructor, ready to be passed into the
+        constructor.
+        """
+        parameters = constructor_signature.parameters
+        assert next(iter(parameters.keys())) == "self"
```

Comment:
why are you checking this?

Reply:
I removed this.  I am keeping the equivalent sanity check for `run` in case we change the way we structure AlgorithmBase fundamentally (e.g., `run` becomes a static or class method).